src/probnum/_config.py

```diff
@@ -36,39 +28,61 @@ class Configuration:
     0.01
     """
 
+    _NON_REGISTERED_KEY_ERR_MSG = (
+        'Configuration option "%s" does not exist yet. '
+        "Configuration options must be `register`ed before they can be "
+        "accessed."
+    )
+
+    @dataclasses.dataclass
+    class Option:
+        name: str
+        default_value: Any
+        description: str
+        value: Any
+
+        def __repr__(self) -> str:
+            _r = "<Configuration.Option "
+            _r += f"name={self.name}, value={self.value}>"
+            return _r
+
+    def __init__(self) -> None:
+        object.__setattr__(self, "_options_registry", dict())
+
+    def __getattr__(self, key: str) -> Any:
+        if key not in self._options_registry:
+            raise AttributeError(f'Configuration option "{key}" does not exist.')
+        return self._options_registry[key].value
+
+    def __setattr__(self, key: str, value: Any) -> None:
+        if key not in self._options_registry:
+            raise AttributeError(Configuration._NON_REGISTERED_KEY_ERR_MSG % key)
+
+        self._options_registry[key].value = value
+
+    def __repr__(self) -> str:
+        return repr(self._options_registry)
+
     @contextlib.contextmanager
     def __call__(self, **kwargs) -> None:
-        old_entries = dict()
+        old_options = dict()
 
         for key, value in kwargs.items():
-            if not hasattr(self, key):
-                raise KeyError(
-                    f"Configuration entry {key} does not exist yet."
-                    "Configuration entries must be `register`ed before they can be "
-                    "accessed."
-                )
+            if key not in self._options_registry:
+                raise AttributeError(Configuration._NON_REGISTERED_KEY_ERR_MSG % key)
 
-            old_entries[key] = getattr(self, key)
+            old_options[key] = self._options_registry[key].value
 
-            setattr(self, key, value)
+            self._options_registry[key].value = value
```

Comment:
so all the options are handled inside the `_options_registry`? Is this attribute introduced somewhere? This is no meaningful review comment, I just don't understand where it is defined (and documented?).

Reply:
`_options_registry` is initialized in the constructor, i.e. via

```
object.__setattr__(self, "_options_registry", dict())
```

This monstrosity is necessary, since we overwrite `__setattr__`, so simply calling

```
self._options_registry = {}
```

sadly doesn't work

Reply:
Okay I didnt see it (maybe because it didnt look like a standard init ;)). Thanks. I will just believe you that it is necessary :D